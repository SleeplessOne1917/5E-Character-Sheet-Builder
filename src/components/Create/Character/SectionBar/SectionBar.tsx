import {
	TouchEvent,
	TouchEventHandler,
	useCallback,
	useEffect,
	useState
} from 'react';

import Link from 'next/link';
import { capitalize } from '../../../../services/capitalizeService';
import classes from './SectionBar.module.css';
import useMediaQuery from '../../../../hooks/useMediaQuery';
import { usePathname } from 'next/navigation';

export enum Section {
	'race',
	'class',
	'abilities',
	'description',
	'equipment',
	'finish'
}

const sections = [
	{ section: 'race', snapPercent: 0 },
	{ section: 'class', snapPercent: 9 },
	{ section: 'abilities', snapPercent: 20 },
	{ section: 'description', snapPercent: 33 },
	{ section: 'equipment', snapPercent: 50 },
	{ section: 'finish', snapPercent: 60 }
];

const spellcastingSections = [
	{ section: 'race', snapPercent: 0 },
	{ section: 'class', snapPercent: 9 },
	{ section: 'abilities', snapPercent: 18 },
	{ section: 'spells', snapPercent: 27 },
	{ section: 'description', snapPercent: 38 },
	{ section: 'equipment', snapPercent: 50 },
	{ section: 'finish', snapPercent: 60 }
];

const getSections = (
	hasSpellcasting: boolean
): { section: string; snapPercent: number }[] =>
	hasSpellcasting ? spellcastingSections : sections;

const MAX_PERCENT = 60;

type SectionBarProps = {
	hasSpellcasting?: boolean;
};

const SectionBar = ({
	hasSpellcasting = false
}: SectionBarProps): JSX.Element => {
	const [translatePercent, setTranslatePercent] = useState(0);
	const [x0, setX0] = useState(0);
	const [selectedSection, setSelectedSection] = useState(
		getSections(hasSpellcasting)[0].section
	);
	const isMediumOrLarger = useMediaQuery('(min-width: 768px)');
	const pathname = usePathname();

	useEffect(() => {
		const sections = getSections(hasSpellcasting);
		const pathRegex = new RegExp(
			String.raw`\/create\/character\/(${sections
				.map(({ section }) => section)
				.join('|')})`
		);
		const match = pathname.match(pathRegex);

		if (match) {
			const key = match[1];
			setSelectedSection(key);
			setTranslatePercent(
				sections.find(section => section.section === key)?.snapPercent as number
			);
		}
	}, [pathname, hasSpellcasting]);

	const getClosestToSnapPercent = useCallback(
		(n: number): number => {
			const diffs = [];
			for (const snapPercent of getSections(hasSpellcasting).map(
				({ snapPercent }) => snapPercent
			)) {
				diffs.push({ diff: Math.abs(n - snapPercent), value: snapPercent });
			}

			diffs.sort((a, b) => {
				if (a.diff < b.diff) {
					return -1;
				} else if (b.diff < a.diff) {
					return 1;
				} else {
					return 0;
				}
			});

			return diffs[0].value;
		},
		[hasSpellcasting]
	);

	const calculatePercent = useCallback(
		(event: TouchEvent<HTMLUListElement>): number => {
			const dx = (event.changedTouches[0].clientX - x0) / 700;

			return Math.max(0, Math.min(translatePercent - dx, MAX_PERCENT));
		},
		[translatePercent, x0]
	);

	const handleSwipe: TouchEventHandler<HTMLUListElement> = useCallback(
		(event: TouchEvent<HTMLUListElement>) => {
			setTranslatePercent(calculatePercent(event));
		},
		[setTranslatePercent, calculatePercent]
	);

	const handleTouchUp: TouchEventHandler<HTMLUListElement> = useCallback(
		(event: TouchEvent<HTMLUListElement>) => {
			const percent = calculatePercent(event);
			setTranslatePercent(getClosestToSnapPercent(percent));
		},
		[setTranslatePercent, getClosestToSnapPercent, calculatePercent]
	);

	return (
		<nav className={classes['section-bar']}>
			<ul
				className={classes['section-list']}
				style={
					isMediumOrLarger
						? {}
						: {
								transform: `translateX(-${translatePercent}%)`
						  }
				}
				onTouchStart={event => setX0(event.changedTouches[0].clientX)}
				onTouchEnd={handleTouchUp}
				onTouchMove={handleSwipe}
				data-testid="section-list"
			>
				{getSections(hasSpellcasting).map(section => (
					<li key={section.section}>
						<Link
							href={`/create/character/${section.section}`}
							passHref
							legacyBehavior
						>
							<a
								className={`${classes.link}${
									section.section === selectedSection
										? ` ${classes['link-selected']}`
										: ''
								}`}
								onClick={() => setTranslatePercent(section.snapPercent)}
								style={
									section.section === selectedSection
										? { transform: 'translateY(0)' }
										: {}
								}
							>
								{capitalize(section.section)}
							</a>
						</Link>
					</li>
				))}
			</ul>
		</nav>
	);
};

export default SectionBar;
