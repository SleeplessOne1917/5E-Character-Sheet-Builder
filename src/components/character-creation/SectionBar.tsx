import { TouchEvent, TouchEventHandler, useState } from "react";

import Link from "next/link";
import classes from "./SectionBar.module.css";

const createSections = {
  race: "Race",
  class: "Class",
  abilities: "Abilities",
  description: "Description",
  equipment: "Equipment",
  finish: "Finish",
};

const SectionBar = (): JSX.Element => {
  const [translatePercent, setTranslatePercent] = useState(0);
  const [x0, setx0] = useState(0);
  let isMediumOrLarger = false;
  if (typeof window !== "undefined") {
    isMediumOrLarger = window?.matchMedia("(min-width: 768px)").matches;
  }

  const setSnapPercent = (index: number) => {
    console.log(index);
    if (index === 3) {
      setTranslatePercent(index * 10 + 6);
    } else if (index >= 4) {
      setTranslatePercent((index + 1) * 10);
    } else {
      setTranslatePercent(index * 10);
    }
  };

  const swipe: TouchEventHandler<HTMLUListElement> = (
    event: TouchEvent<HTMLUListElement>
  ) => {
    const dx = -(event.changedTouches[0].clientX - x0);

    const percent = Math.max(0, Math.min(dx, 60));
    setTranslatePercent(percent);
  };

  return (
    <nav className={classes["section-bar"]}>
      <ul
        className={classes["section-list"]}
        style={
          isMediumOrLarger
            ? {}
            : {
                transform: `translateX(-${translatePercent}%)`,
              }
        }
        onTouchStart={(event) => setx0(event.changedTouches[0].clientX)}
        onTouchEnd={swipe}
        onTouchMove={swipe}
      >
        {Object.entries(createSections).map(([key, value], index) => (
          <li key={key}>
            <Link href={`/create/race#`}>
              <a className={classes.link} onClick={() => setSnapPercent(index)}>
                {value}
              </a>
            </Link>
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default SectionBar;
