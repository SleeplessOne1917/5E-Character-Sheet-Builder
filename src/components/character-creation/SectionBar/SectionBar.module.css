.section-bar {
  background-color: rgba(29, 0, 0, 0.8);
  z-index: 10;
  overflow-x: hidden;
  min-height: 5.3rem;
}

.section-list {
  padding: 0.5rem;
  display: flex;
  justify-content: space-around;
  align-items: center;
  width: 270%;
  list-style: none;
  transition: transform 0.5s;
}

.section-list > li:first-of-type {
  margin-left: 8rem;
}

.section-list > li:last-of-type {
  margin-right: 10rem;
}

.link {
  display: block;
  color: var(--heaven);
  font-family: var(--font-fantasy);
  font-size: 1rem;
  line-height: 1;
  text-decoration: none;
  width: fit-content;
  padding: 0.25rem;
  z-index: 20;
}

.link:focus-visible {
  outline: var(--focus-outline);
}

.link-selected {
  color: var(--highlight-active);
  border-bottom: 5px solid var(--highlight-active);
}

@media screen and (min-width: 768px) {
  .section-list {
    width: 100%;
    justify-content: center;
  }

  .link {
    margin: 0 1.5rem;
    font-size: 1.3rem;
  }
}

@media screen and (min-width: 1024px) {
  .link {
    margin: 0 2rem;
    font-size: 1.5rem;
  }

  .link:hover {
    color: var(--highlight);
    transform: translateY(-0.25rem);
  }
}
