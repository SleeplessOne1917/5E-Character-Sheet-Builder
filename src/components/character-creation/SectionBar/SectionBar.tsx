import { TouchEvent, TouchEventHandler, useEffect, useState } from 'react';

import Link from 'next/link';
import classes from './SectionBar.module.css';
import useMediaQuery from '../../../hooks/useMediaQuery';
import { useRouter } from 'next/router';

export enum Section {
	'race',
	'class',
	'abilities',
	'description',
	'equipment',
	'finish'
}

const getSections = (): string[] =>
	Object.keys(Section).filter(key => isNaN(Number(key)));

const snapPercents = [0, 9, 20, 33, 50, 60];

const MAX_PERCENT = 60;

const capitalize = (str: string) => {
	const [first, ...rest] = str;
	return first.toUpperCase() + rest.join('').toLowerCase();
};

const getClosestToSnapPercent = (n: number): number => {
	const diffs = [];
	for (const snapPercent of snapPercents) {
		diffs.push({ diff: Math.abs(n - snapPercent), value: snapPercent });
	}

	diffs.sort((a, b) => {
		if (a.diff < b.diff) {
			return -1;
		} else if (b.diff < a.diff) {
			return 1;
		} else {
			return 0;
		}
	});

	return diffs[0].value;
};

const SectionBar = (): JSX.Element => {
	const [translatePercent, setTranslatePercent] = useState(0);
	const [x0, setX0] = useState(0);
	const [selectedSection, setSelectedSection] = useState(Section[0]);
	const { pathname } = useRouter();
	const isMediumOrLarger = useMediaQuery('(min-width: 768px)');

	useEffect(() => {
		const pathRegex = new RegExp(
			String.raw`\/create\/(${getSections().join('|')})`
		);
		const match = pathname.match(pathRegex);

		if (match) {
			const key = match[1];
			setSelectedSection(key);
			setTranslatePercent(snapPercents[Section[key]]);
		}
	}, [pathname]);

	const calculatePercent = (event: TouchEvent<HTMLUListElement>): number => {
		const dx = (event.changedTouches[0].clientX - x0) / 700;

		return Math.max(0, Math.min(translatePercent - dx, MAX_PERCENT));
	};

	const handleSwipe: TouchEventHandler<HTMLUListElement> = (
		event: TouchEvent<HTMLUListElement>
	) => {
		setTranslatePercent(calculatePercent(event));
	};

	const handleTouchUp: TouchEventHandler<HTMLUListElement> = (
		event: TouchEvent<HTMLUListElement>
	) => {
		const percent = calculatePercent(event);
		setTranslatePercent(getClosestToSnapPercent(percent));
	};

	return (
		<nav className={classes['section-bar']}>
			<ul
				className={classes['section-list']}
				style={
					isMediumOrLarger
						? {}
						: {
								transform: `translateX(-${translatePercent}%)`
						  }
				}
				onTouchStart={event => setX0(event.changedTouches[0].clientX)}
				onTouchEnd={handleTouchUp}
				onTouchMove={handleSwipe}
				data-testid="section-list"
			>
				{getSections().map((key, index) => (
					<li key={key}>
						<Link href={`/create/${key}`}>
							<a
								className={`${classes.link}${
									key === selectedSection ? ` ${classes['link-selected']}` : ''
								}`}
								onClick={() => setTranslatePercent(snapPercents[index])}
								style={
									key === selectedSection ? { transform: 'translateY(0)' } : {}
								}
							>
								{capitalize(key)}
							</a>
						</Link>
					</li>
				))}
			</ul>
		</nav>
	);
};

export default SectionBar;
