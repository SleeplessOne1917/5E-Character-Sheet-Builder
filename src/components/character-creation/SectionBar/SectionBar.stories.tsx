import { ComponentMeta, ComponentStory } from '@storybook/react';

import SectionBar from './SectionBar';

export default {
	title: 'Components/SectionBar',
	component: SectionBar
} as ComponentMeta<typeof SectionBar>;

export const Default: ComponentStory<typeof SectionBar> = () => <SectionBar />;
Default.parameters = {
	nextRouter: {
		pathname: '/create/race',
		asPath: '/create/race'
	}
};
