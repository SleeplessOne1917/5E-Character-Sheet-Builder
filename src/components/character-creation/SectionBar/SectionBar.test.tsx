import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';

import SectionBar from './SectionBar';

const mockRouter = (pathname: string) => {
	const useRouter = jest.spyOn(require('next/router'), 'useRouter');
	const router = {
		push: () => Promise.resolve(),
		prefetch: () => Promise.resolve(),
		pathname
	};
	useRouter.mockReturnValue(router);
};

const mockMatchMedia = (matches: boolean) => {
	window.matchMedia = jest.fn().mockImplementationOnce(() => ({
		matches,
		addEventListener: () => {},
		removeEventListener: () => {}
	}));
};

it('renders correctly', () => {
	mockMatchMedia(true);
	mockRouter('/create/race/');

	render(<SectionBar />);

	expect(screen.getByRole('navigation')).toMatchSnapshot();
});

describe('snaps to snapPercents', () => {
	beforeEach(() => {
		mockMatchMedia(false);
	});

	it('at race', () => {
		mockRouter('/create/race/');
		render(<SectionBar />);

		expect(screen.getByTestId(/section-list/i)).toHaveStyle(
			'transform: translateX(-0%)'
		);
	});

	it('at class', () => {
		mockRouter('/create/class/');
		render(<SectionBar />);

		expect(screen.getByTestId(/section-list/i)).toHaveStyle(
			'transform: translateX(-9%)'
		);
	});

	it('at abilities', () => {
		mockRouter('/create/abilities/');
		render(<SectionBar />);

		expect(screen.getByTestId(/section-list/i)).toHaveStyle(
			'transform: translateX(-20%)'
		);
	});

	it('at description', () => {
		mockRouter('/create/description/');
		render(<SectionBar />);

		expect(screen.getByTestId(/section-list/i)).toHaveStyle(
			'transform: translateX(-33%)'
		);
	});

	it('at equipment', () => {
		mockRouter('/create/equipment/');
		render(<SectionBar />);

		expect(screen.getByTestId(/section-list/i)).toHaveStyle(
			'transform: translateX(-50%)'
		);
	});

	it('at finish', () => {
		mockRouter('/create/finish/');
		render(<SectionBar />);

		expect(screen.getByTestId(/section-list/i)).toHaveStyle(
			'transform: translateX(-60%)'
		);
	});
});
