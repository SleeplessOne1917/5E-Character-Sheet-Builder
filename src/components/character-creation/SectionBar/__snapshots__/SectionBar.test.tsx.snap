// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<nav
  class="section-bar"
>
  <ul
    class="section-list"
    data-testid="section-list"
    style=""
  >
    <li>
      <a
        class="link link-selected"
        href="/create/race"
        style="transform: translateY(0);"
      >
        Race
      </a>
    </li>
    <li>
      <a
        class="link"
        href="/create/class"
      >
        Class
      </a>
    </li>
    <li>
      <a
        class="link"
        href="/create/abilities"
      >
        Abilities
      </a>
    </li>
    <li>
      <a
        class="link"
        href="/create/description"
      >
        Description
      </a>
    </li>
    <li>
      <a
        class="link"
        href="/create/equipment"
      >
        Equipment
      </a>
    </li>
    <li>
      <a
        class="link"
        href="/create/finish"
      >
        Finish
      </a>
    </li>
  </ul>
</nav>
`;
